### PR TITLE
Fix for 1651

### PR DESCRIPTION
**Pull request by sbrossie**
This PR implements the support for setting a `quantity` when creating a subscription: Setting a quantity=3 is equivalent to creating 3 times the same subscription (i.e any FIXED or RECURRING items would be billed 3 times their normal price). See #1651 

This is rather large PR, so looking at individual commits may make sense. The work is mostly at the `subscription` level and the implementation is very well aligned with the one we have for BCD support. New apis and bus events have been created as a result. This PR depends on:

* https://github.com/killbill/killbill-plugin-api/pull/29
* https://github.com/killbill/killbill-api/pull/112
* https://github.com/killbill/killbill-client-java/pull/116

### CODE REVIEW

File: subscription/src/main/java/org/killbill/billing/subscription/api/transfer/DefaultSubscriptionBaseTransferApi.java [hunk 1]
```diff
@@ -120,6 +120,10 @@ private SubscriptionBaseEvent createEvent(final boolean firstEvent, final Existi
                 newEvent = BCDEventData.createBCDEvent(subscription, effectiveDate, existingEvent.getBillCycleDayLocal());
                 break;
 
+            case QUANTITY_CHANGE:
+                // TODO
```
Comment (pierre) [inline subscription/src/main/java/org/killbill/billing/subscription/api/transfer/DefaultSubscriptionBaseTransferApi.java:124]: Maybe throw unsupported?

File: subscription/src/main/java/org/killbill/billing/subscription/api/user/DefaultSubscriptionBase.java [hunk 2]
```diff
@@ -460,14 +460,26 @@ public ProductCategory getCategory() {
...
+        return getQuantityOrBCD(true);
+    }
+
+    @Override
+    public Integer getQuantity() {
+        return getQuantityOrBCD(false);
+    }
+
+
+    private Integer getQuantityOrBCD(final boolean bcd) {
```
Comment (pierre) [inline subscription/src/main/java/org/killbill/billing/subscription/api/user/DefaultSubscriptionBase.java:472]: Nit: I find the method a tad confusing.

Maybe `getNextValueForTransition(final SubscriptionBaseTransitionType transitionType)`?

File: invoice/src/main/resources/org/killbill/billing/invoice/ddl.sql [hunk 3]
```diff
@@ -106,7 +106,7 @@ CREATE TABLE invoice_item_history (
     rate numeric(15,9) NULL,
     currency varchar(3) NOT NULL,
     linked_item_id varchar(36),
-    quantity int,
+    quantity decimal(18, 9),
```
Comment (pierre) [inline invoice/src/main/resources/org/killbill/billing/invoice/ddl.sql:109]: Shouldn't this be symmetric with subscriptions?

**Review by pierre — APPROVED** [review]

File: subscription/src/main/java/org/killbill/billing/subscription/api/transfer/DefaultSubscriptionBaseTransferApi.java (hunk 1 above)
Comment (sbrossie) [reply subscription/src/main/java/org/killbill/billing/subscription/api/transfer/DefaultSubscriptionBaseTransferApi.java:124]: Just to be clear: With the current code we allow the transfer, but we lose the `quantity` as it is not 'ported' with the subscription. If I understand correctly you are proposing the prevent the transfer from happening in such case to make it clear?

Ideally, I would like to fix it in as part of #1627, which I kept opened.

File: invoice/src/main/resources/org/killbill/billing/invoice/ddl.sql (hunk 3 above)
Comment (sbrossie) [reply invoice/src/main/resources/org/killbill/billing/invoice/ddl.sql:109]: That was my thinking as well, but as I was looking for what we **already have** (in invoice), I discovered our scheme is inconsistent: [invoice_items](https://github.com/killbill/killbill/blob/work-for-release-0.23.x/invoice/src/main/resources/org/killbill/billing/invoice/ddl.sql#L71) versus [invoice_items_history](https://github.com/killbill/killbill/blob/work-for-release-0.23.x/invoice/src/main/resources/org/killbill/billing/invoice/ddl.sql#L109)

File: subscription/src/main/java/org/killbill/billing/subscription/api/user/DefaultSubscriptionBase.java (hunk 2 above)
Comment (sbrossie) [reply subscription/src/main/java/org/killbill/billing/subscription/api/user/DefaultSubscriptionBase.java]: Fixed in https://github.com/killbill/killbill/pull/1805/commits/9e05be1dc910b224d8bc28994c7d43a909225277

File: subscription/src/main/java/org/killbill/billing/subscription/api/transfer/DefaultSubscriptionBaseTransferApi.java (hunk 1 above)
Comment (sbrossie) [reply subscription/src/main/java/org/killbill/billing/subscription/api/transfer/DefaultSubscriptionBaseTransferApi.java]: Followed your recommendation for now in https://github.com/killbill/killbill/pull/1805/commits/e375c3c360a421c90cebff5e262d5d9f7ca9c490

**Pull request merged**